### PR TITLE
send leader term with heart and reorganize leader table

### PR DESCRIPTION
**Pull request by liuyu85cn**
For short, we use term as Lamport Clock to avoid some heavy version check in BASE TOSS insertion. 

related common PR(has to merge together):
https://github.com/vesoft-inc/nebula-common/pull/485

### CODE REVIEW

File: src/kvstore/raftex/RaftPart.h [hunk 1]
```diff
@@ -127,6 +127,10 @@ class RaftPart : public std::enable_shared_from_this<RaftPart> {
         return leader_;
     }
 
+    TermID termId() const {
+        return term_;
```
Comment (critical27) [inline src/kvstore/raftex/RaftPart.h:131]: Grab the lock?

File: src/meta/MetaServiceUtils.cpp [hunk 2]
```diff
@@ -280,7 +285,17 @@ std::string MetaServiceUtils::leaderKeyV2(std::string addr, Port port) {
...
+    std::string key;
+    key.reserve(kLeaderTermsTable.size() + sizeof(GraphSpaceID) + sizeof(PartitionID));
+    key.append(kLeaderTermsTable.data(), kLeaderTermsTable.size())
+        .append(reinterpret_cast<const char*>(&spaceId), sizeof(GraphSpaceID))
+        .append(reinterpret_cast<const char*>(&partId), sizeof(PartitionID));
+    return key;
+}
+
 std::string MetaServiceUtils::leaderVal(const LeaderParts& leaderParts) {
+    LOG(ERROR) << "deprecated function\n" << boost::stacktrace::stacktrace();
```
Comment (critical27) [inline src/meta/MetaServiceUtils.cpp:298]: Maybe we could ignore the leader key when upgrade meta in `MetaVersionMan`

File: src/meta/MetaServiceUtils.cpp [hunk 3]
```diff
@@ -295,11 +310,49 @@ std::string MetaServiceUtils::leaderVal(const LeaderParts& leaderParts) {
...
+
+    leaderVal.append(reinterpret_cast<const char*>(&dataVersion), sizeof(dataVersion))
+             .append(reinterpret_cast<const char*>(&lenOfHost), sizeof(lenOfHost))
+             .append(sHost)
+             .append(reinterpret_cast<const char*>(&term), sizeof(term));
+    return leaderVal;
+}
+
+// v3: dataVer(int) + lenOfHost(8) + HostAddr(varchar) + term(int64_t)
+std::pair<HostAddr, int64_t> MetaServiceUtils::parseLeaderValV3(folly::StringPiece val) {
```
Comment (critical27) [inline src/meta/MetaServiceUtils.cpp:332]: Maybe we need to consider backward compatibility here, return a Status or sth else. And handle it in `ActiveHostsMan::updateHostInfo`

**Review by critical27 — APPROVED** [review]
LGTM now

File: src/meta/ActiveHostsMan.cpp [hunk 4]
```diff
@@ -17,15 +18,42 @@ namespace meta {
...
+                if (terms[i] <= std::get<1>(hostAndTerm)) {
+                    continue;
+                }
+            }
+            // write directly if not exist, or update if has greater term
+            auto val = MetaServiceUtils::leaderValV3(hostAddr, terms[i]);
+            data.emplace_back(std::make_pair(leaderKeys[i], std::move(val)));
+        }
     }
+    std::vector<std::string> leaderVals(leaderKeys.size());
```
Comment (panda-sheep) [inline src/meta/ActiveHostsMan.cpp:56]: not used?

File: src/meta/ActiveHostsMan.cpp [hunk 5]
```diff
@@ -17,15 +18,42 @@ namespace meta {
...
+            auto spaceId = spaceLeaders.first;
+            for (auto& partLeader : spaceLeaders.second) {
+                auto key = MetaServiceUtils::leaderKey(spaceId, partLeader.get_part_id());
+                leaderKeys.emplace_back(std::move(key));
+                terms.emplace_back(partLeader.get_term());
+            }
+        }
+        auto keys = leaderKeys;
+        std::vector<std::string> values;
+        auto ret = kv->multiGet(kDefaultSpaceId, kDefaultPartId, std::move(keys), &values);
```
Comment (panda-sheep) [inline src/meta/ActiveHostsMan.cpp:39]: Here you need to judge whether `ret.first == kvstore::ResultCode::SUCCEEDED`
If they are equal, do the following

File: src/kvstore/NebulaStore.cpp [hunk 6]
```diff
@@ -1008,16 +1008,19 @@ ErrorOr<ResultCode, std::shared_ptr<SpacePartInfo>> NebulaStore::space(GraphSpac
...
     for (const auto& spaceIt : spaces_) {
         auto spaceId = spaceIt.first;
         for (const auto& partIt : spaceIt.second->parts_) {
             auto partId = partIt.first;
             if (partIt.second->isLeader()) {
-                leaderIds[spaceId].emplace_back(partId);
+                meta::cpp2::LeaderInfo partInfo;
+                partInfo.set_part_id(partId);
+                partInfo.set_term(partIt.second->termId());
+                leaderIds[spaceId].emplace_back(partInfo);
```
Comment (monadbobo) [inline src/kvstore/NebulaStore.cpp:1023]: std::move(partInfo)

File: src/meta/MetaServiceUtils.cpp [hunk 7]
```diff
@@ -295,11 +310,55 @@ std::string MetaServiceUtils::leaderVal(const LeaderParts& leaderParts) {
...
+    leaderVal.append(reinterpret_cast<const char*>(&dataVersion), sizeof(dataVersion))
+             .append(reinterpret_cast<const char*>(&lenOfHost), sizeof(lenOfHost))
+             .append(sHost)
+             .append(reinterpret_cast<const char*>(&term), sizeof(term));
+    return leaderVal;
+}
+
+// v3: dataVer(int) + lenOfHost(8) + HostAddr(varchar) + term(int64_t)
+std::tuple<HostAddr, int64_t, cpp2::ErrorCode>
+MetaServiceUtils::parseLeaderValV3(folly::StringPiece val) {
```
Comment (monadbobo) [inline src/meta/MetaServiceUtils.cpp:333]: maybe use ErrorOr<cpp2::ErrorCode, std::tuple<HostAddr, int64_t> was better.

File: src/meta/ActiveHostsMan.cpp (hunk 4 above)
Comment (liuyu85cn) [reply src/meta/ActiveHostsMan.cpp]: good catch, fixed

File: src/meta/ActiveHostsMan.cpp (hunk 5 above)
Comment (liuyu85cn) [reply src/meta/ActiveHostsMan.cpp]: good catch, fixed

File: src/kvstore/NebulaStore.cpp (hunk 6 above)
Comment (liuyu85cn) [reply src/kvstore/NebulaStore.cpp]: good catch, fixed

File: src/meta/MetaServiceUtils.cpp (hunk 7 above)
Comment (liuyu85cn) [reply src/meta/MetaServiceUtils.cpp:332]: I prefer tuple

File: src/meta/MetaServiceUtils.cpp [hunk 8]
```diff
@@ -295,11 +310,53 @@ std::string MetaServiceUtils::leaderVal(const LeaderParts& leaderParts) {
...
+    CHECK_GE(val.size(), sizeof(int));
+    val.advance(sizeof(int));
+    auto lenOfHost = *reinterpret_cast<const size_t*>(val.data());
+
+    val.advance(sizeof(size_t));
+    CHECK_GE(val.size(), lenOfHost);
+    std::get<0>(ret) = MetaServiceUtils::deserializeHostAddr(val.subpiece(0, lenOfHost));
+
+    val.advance(lenOfHost);
+    std::get<1>(ret) = *reinterpret_cast<const GraphSpaceID*>(val.data());
```
Comment (panda-sheep) [inline src/meta/MetaServiceUtils.cpp:350]: why GraphSpaceID ?  Shouldn't it be TermID?

**Review by panda-sheep — COMMENTED** [review]
Good job！

File: src/meta/MetaServiceUtils.cpp [hunk 9]
```diff
@@ -318,11 +376,21 @@ HostAddr MetaServiceUtils::parseLeaderKeyV1(folly::StringPiece key) {
 }
 
 HostAddr MetaServiceUtils::parseLeaderKeyV2(folly::StringPiece key) {
+    LOG(FATAL) << "deprecated function";
```
Comment (panda-sheep) [inline src/meta/MetaServiceUtils.cpp:379]: There is a problem here,
 if upgrade. Is the previous meta leader data invalid?
How to delete?

**Review by critical27 — DISMISSED** [review]
LGTM now

**Review by panda-sheep — APPROVED** [review]

**Review by bright-starry-sky — APPROVED** [review]
LGTM

**Comment by bright-starry-sky** [comment]
The CI failed on system `ubuntu1804, gcc-9.2` , it seems an OOM error.
Ignoring that for the moment, and PR #431 will solve this problem.

**Pull request merged**